staker/bold/bold_state_provider.go

```diff
@@ -0,0 +1,539 @@
+	var prevBatchMsgCount arbutil.MessageIndex
+	bNum := md.FromState.Batch
+	posInBatch := md.FromState.PosInBatch
+	if bNum > 0 {
+		var err error
+		prevBatchMsgCount, err = s.statelessValidator.InboxTracker().GetBatchMessageCount(uint64(bNum - 1))
+		if err != nil {
+			return 0, fmt.Errorf("could not get prevBatchMsgCount at %d: %w", bNum-1, err)
+		}
+	}
+	return prevBatchMsgCount + arbutil.MessageIndex(posInBatch) + arbutil.MessageIndex(chalHeight), nil
+}
+
+// virtualState returns an optional global state.
+//
+// If messageNum is a virtual block or the last real block to which this
+// validator's assertion committed, then this function retuns a global state
+// representing that virtual block's finished machine. Otherwise, it returns
+// an Option.None.
+//
+// This can happen in the BoLD protocol when the rival block-level challenge
+// edge has committed to more blocks that this validator expected for the
+// current batch. In that case, the chalHeight will be a block in the virtual
+// padding of the history commitment of this validator.
+//
+// If there is an Option.Some() retrun value, it means that callers don't need
+// to actually step through a machine to produce a series of hashes, because all
+// of the hashes can just be "virtual" copies of a single machine in the
+// FINISHED state's hash.
+func (s *BOLDStateProvider) virtualState(msgNum arbutil.MessageIndex, limit l2stateprovider.Batch) (option.Option[validator.GoGlobalState], error) {
+	gs := option.None[validator.GoGlobalState]()
+	limitMsgCount, err := s.statelessValidator.InboxTracker().GetBatchMessageCount(uint64(limit) - 1)
+	if err != nil {
+		return gs, fmt.Errorf("could not get limitMsgCount at %d: %w", limit, err)
+	}
+	if msgNum >= limitMsgCount {
+		result, err := s.statelessValidator.InboxStreamer().ResultAtCount(arbutil.MessageIndex(limitMsgCount))
+		if err != nil {
+			return gs, fmt.Errorf("could not get global state at limitMsgCount %d: %w", limitMsgCount, err)
+		}
+		gs = option.Some(validator.GoGlobalState{
+			BlockHash:  result.BlockHash,
+			SendRoot:   result.SendRoot,
+			Batch:      uint64(limit),
+			PosInBatch: 0,
+		})
+	}
+	return gs, nil
+}
+
+// CtxWithCheckAlive Creates a context with a check alive routine that will
+// cancel the context if the check alive routine fails.
+func ctxWithCheckAlive(ctxIn context.Context, execRun validator.ExecutionRun) (context.Context, context.CancelFunc) {
+	// Create a context that will cancel if the check alive routine fails.
+	// This is to ensure that we do not have the validator froze indefinitely if
+	// the execution run is no longer alive.
+	ctx, cancel := context.WithCancel(ctxIn)
+	// Create a context with cancel, so that we can cancel the check alive routine
+	// once the calling function returns.
+	ctxCheckAlive, cancelCheckAlive := context.WithCancel(ctxIn)
```

Comment:
Couldn't ctxCheckAlive be the same as ctx? Not that it matters much, because as soon as ctxCheckAlive is canceled the defer in the background goroutine will also cancel ctx.